### PR TITLE
[WIP] Expose API that allows for network resizing images

### PR DESCRIPTION
**Pull request by iskakaushik**
## Description

This PR creates a new image provider `ResizedImage` that allows the user to specify the dimensions of the image that the engine can then use to resize to. This is the engine side PR accounting for the same changes: https://github.com/flutter/engine/pull/8596

## Related Issues

https://github.com/flutter/flutter/issues/26194

## Tests

NEEDS TESTS!!

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [x] No, this is *not* a breaking change.

### CODE REVIEW

File: packages/flutter/lib/src/painting/image_provider.dart [hunk 1]
```diff
@@ -460,11 +471,48 @@ abstract class AssetBundleImageProvider extends ImageProvider<AssetBundleImageKe
...
   /// This function is used by [load].
   @protected
-  Future<ui.Codec> _loadAsync(AssetBundleImageKey key) async {
+  Future<ui.Codec> _loadAsync(AssetBundleImageKey key, {Size size}) async {
     final ByteData data = await key.bundle.load(key.name);
     if (data == null)
       throw 'Unable to read data';
-    return await PaintingBinding.instance.instantiateImageCodec(data.buffer.asUint8List());
+    return await PaintingBinding.instance.instantiateImageCodec(
+        data.buffer.asUint8List(), size: size);
```
Comment (jonahwilliams) [inline packages/flutter/lib/src/painting/image_provider.dart]: One line

File: packages/flutter/lib/src/painting/image_provider.dart [hunk 2]
```diff
@@ -460,11 +471,48 @@ abstract class AssetBundleImageProvider extends ImageProvider<AssetBundleImageKe
...
+
+  @override
+  ImageStreamCompleter load(_SizeAwareCacheKey key, {Size size}) {
+    assert(key._imageSize == _imageSize);
+    return _imageProvider.load(key._providerCacheKey, size: _imageSize);
+  }
+
+  @override
+  Future<_SizeAwareCacheKey> obtainKey(
+      ImageConfiguration configuration) async {
```
Comment (jonahwilliams) [inline packages/flutter/lib/src/painting/image_provider.dart]: The ImageConfiguration already has the size information, we should find a way to use that

File: packages/flutter/lib/src/widgets/image.dart [hunk 3]
```diff
@@ -197,6 +214,13 @@ class Image extends StatefulWidget {
   /// [FilterQuality.none] which corresponds to nearest-neighbor.
   ///
   /// If [excludeFromSemantics] is true, then [semanticLabel] will be ignored.
+  ///
+  /// If [resizeToFit] parameter is set to true, [width] and [height] must be
+  /// set for resizing to take effect. This parameter indicates to the engine
+  /// that this image must be resized. The image will be sized to the constraints
+  /// of the layout or [width] and [height] regardless of this parameter. This
+  /// parameter is primarily intended for use to reduce the memory usage of
```
Comment (jonahwilliams) [inline packages/flutter/lib/src/widgets/image.dart]:
```suggestion
  /// parameter is primarily intended to reduce the memory usage of
```

File: packages/flutter/lib/src/widgets/image.dart [hunk 4]
```diff
@@ -197,6 +214,13 @@ class Image extends StatefulWidget {
...
+  /// [ImageCache].
```
Comment (jonahwilliams) [inline packages/flutter/lib/src/widgets/image.dart]: We should gather some numbers to confirm this is true.

File: packages/flutter/lib/src/painting/image_provider.dart [hunk 5]
```diff
@@ -460,11 +471,48 @@ abstract class AssetBundleImageProvider extends ImageProvider<AssetBundleImageKe
...
+  final Size _imageSize;
+}
+
+/// Re-sizes the image provided to the given size.
+///
+/// See also:
+///
+/// * [Image.network] for example usage when `resizeToFit` parameter is set.
+class ResizedImage
+    extends ImageProvider<_SizeAwareCacheKey> {
```
Comment (Hixie) [inline packages/flutter/lib/src/painting/image_provider.dart]: nit: no need for newline here

File: packages/flutter/lib/src/widgets/image.dart [hunk 6]
```diff
@@ -215,7 +239,15 @@ class Image extends StatefulWidget {
     this.gaplessPlayback = false,
     this.filterQuality = FilterQuality.low,
     Map<String, String> headers,
-  }) : image = NetworkImage(src, scale: scale, headers: headers),
+    bool resizeToFit = false,
+  }) : image = _resizeIfNeeded(
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/image.dart]: why do we do this for network image but not asset image or memory image?

**Comment by goderbauer** [comment]
@iskakaushik are you still working on this one or should we close it?

**Comment by iskakaushik** [comment]
> @iskakaushik are you still working on this one or should we close it?

@goderbauer yes. I am waiting on the engine side to reach a good state first.

File: packages/flutter/lib/src/widgets/image.dart (hunk 6 above)
Comment (iskakaushik) [reply packages/flutter/lib/src/widgets/image.dart]: @Hixie , i wanted to get feedback on this `_resizeIfNeeded` pattern before i go ahead and change the other ones.

Comment (Hixie) [reply packages/flutter/lib/src/widgets/image.dart]: I don't really understand the pattern as written. It seems very network-specific. I'm having trouble understanding how it generalizes.

**Comment by jonahwilliams** [comment]
We should be able to resize any image with this, but FileImage would be another usecase outlined in the original doc:

- Take high resolution picture with phone camera
- Save to filesystem
- Load preview image in flutter app that is much smaller

**Comment by iskakaushik** [comment]
@jonahwilliams I added support for memory, file and network images with my latest changes. Also added some tests -- though these only test memory images for now. I will add some more to account for network and memory images.

**Comment by goderbauer** [comment]
@iskakaushik Do you still have plans for this PR? It is getting a little stale. :)

**Comment by iskakaushik** [comment]
closing this for now. will reopen after adding some benchmarks for current image decoding.

**Pull request closed without merging**